docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,402 @@
+A function's return type and the types of parameters may be omitted;
+Relay will attempt to infer the most general types where types
+are omitted.
+
+Functions defined in the manner described in this subsection are
+of global scope; anonymous functions will be discussed later, though
+their mechanics are nearly identical. Note that global functions may be
+recursive; that is, within the function's body, the function's
+identifier refers back to the function unless it is shadowed in a `let`
+expression.
+
+A definition minimally consists of an identifier :code:`@id`, an empty set of
+parameters, and a body expression (:py:class:`~tvm.relay.expr.Expr`)
+contained by curly braces.
+
+.. code-block:: python
+
+    def @id() { body }
+
+A definition may contain any number of parameters. For example, a
+simple function that invokes the `add` operator:
+
+.. code-block:: python
+
+    def @plus(%x, %y) { add(%x, %y) }
+
+Notice that within the function's body, the parameters are local
+variables, just like those bound in a `let` expression.
+
+It is also possible for us to annotate explicit types on definitions.
+For example, we can restrict the above definition to only work
+on certain types:
+
+.. code-block:: python
+
+    def @add(%x: Tensor<Float, (10, 10)>, %y: Tensor<Float, (10, 10)>) -> Tensor<Float, (10, 10)> {
+        %x + %y
+    }
+
+A parameter is just a pairing of a :py:class:`~tvm.relay.expr.LocalVar` and optional :py:class:`~tvm.relay.ty.Type`. They represent
+the formal parameters of functions and definitions and are written as :code:`%x : T`.
+
+Parameters may only appear in function literals and definitions
+and have no relation to parameters in the machine learning.
+
+When the type information is omitted, we will attempt to infer the most general type
+for the users. This property is known as generalization: for a definition without
+explicit annotations, we will attempt to assign the most general type. When the
+return type is omitted, we will infer the return type based on the text of the
+program.
+
+We can directly construct type-polymorphic definitions by writing down
+a set of type parameters for a definition. For example, one can define a
+polymorphic identity function for tensors as follows:
+
+.. code-block:: python
+    def @id<s: Shape, bt: BaseType>(%x: Tensor<bt, s>) {
+        %x
+    }
+
+Notice that the return type is omitted and will be inferred.
+
+.. *Note: polymorphism is not yet implemented.*
+
+A function may also be subject to one or more type relations, such as in
+the following:
+
+.. code-block:: python
+
+    def @plus(%x, %y) where Broadcast { add(%x, %y) }
+
+In the above definition, the types of `%x` and `%y` and the return type
+are subject to the `Broadcast` relation, meaning all three must be tensors
+and their shapes follow the elementwise broadcast relation. As with
+operators, the definitions of relations are not transparent to Relay
+and they are instead implemented externally.
```

Comment:
I think you should be specific here, i.e C++ or Python

Reply:
Done in e7c5a04